File: Prism/Settings/AppSettings.swift
import Foundation
import ServiceManagement

struct AppSettings: Equatable {
    var isEnabled: Bool = true
    var enterTranslatesAndSends: Bool = false

    // MARK: Idiomas — mensagens recebidas (só leitura / popup de leitura)

    /// nil = detecção automática.
    var incomingSourceLanguage: String? = nil
    /// Idioma para o qual traduzir o que você lê (padrão = idioma do sistema).
    var incomingTargetLanguage: String = LanguageCode.systemLanguage

    // MARK: Idiomas — suas mensagens (campo editável / replace)

    /// nil = detecção automática; padrão = idioma do sistema.
    var outgoingSourceLanguage: String? = LanguageCode.systemLanguage
    /// Idioma para o qual traduzir o que você escreve (padrão = inglês, ou pt se o sistema for en).
    var outgoingTargetLanguage: String = LanguageCode.systemDefaultTarget

    var providerKind: ProviderKind = .apple
    var openAtLogin: Bool = false

    /// ⌃⌥T by default — avoids ⌘ conflicts with browsers and system apps.
    var translateOnlyHotkey: HotkeyChord = .translateOnlyDefault
    /// ⌃⌥⏎ by default — avoids ⌥⏎ (newline / IDE quick-fix).
    var translateAndSendHotkey: HotkeyChord = .translateAndSendDefault
    /// Dedicated popup mode (selection → panel). Off by default (opt-in).
    var popupModeEnabled: Bool = false
    /// ⌃⌥Y by default — show translation popup.
    var popupHotkey: HotkeyChord = .popupDefault
    /// Toast perto do ponteiro (Traduzindo / concluído / erro) ao usar o atalho.
    var showStatusHUD: Bool = true

    // DeepL
    var deeplUseFreeAPI: Bool = true

    func model(for kind: ProviderKind) -> String? {
        switch kind {
        case .openAICompatible: return openAIModel
        default: return nil
        }
    }

    mutating func setModel(_ model: String, for kind: ProviderKind) {
        switch kind {
        case .openAICompatible: openAIModel = model
        default: break
        }
    }

    // OpenAI-compatible / LM Studio
    var openAIBaseURL: String = "http://localhost:1234/v1"
    var openAIModel: String = "local-model"

    // Custom HTTP
    var customHTTPURL: String = "http://localhost:8080/translate"
    var customHTTPMethod: String = "POST"
    var customHTTPHeadersJSON: String = #"{"Content-Type":"application/json"}"#
    var customHTTPBodyTemplate: String = #"{"text":"{{text}}","target":"{{to}}","source":"{{from}}"}"#
    var customHTTPResponsePath: String = "translatedText"

    /// Source to send to providers for a direction; nil means auto-detect.
    func resolvedSourceLanguage(outgoing: Bool) -> String? {
        let raw = outgoing ? outgoingSourceLanguage : incomingSourceLanguage
        guard let raw, !raw.isEmpty else { return nil }
        return raw
    }

    func targetLanguage(outgoing: Bool) -> String {
        outgoing ? outgoingTargetLanguage : incomingTargetLanguage
    }

    /// Convenience for call sites that still think in “your messages” terms.
    var resolvedOutgoingSourceLanguage: String? {
        resolvedSourceLanguage(outgoing: true)
    }

    /// Nome do motor + modelo/endpoint — uma linha fácil de ler no log.
    var engineLogDescription: String {
        let name = providerKind.displayName
        switch providerKind {
        case .apple:
            return "\(name) (no dispositivo)"
        case .deepl:
            return "\(name) · \(deeplUseFreeAPI ? "API gratuita" : "API Pro")"
        case .google:
            return name
        case .openAICompatible:
            return "\(name) · modelo \(openAIModel) · \(openAIBaseURL)"
        case .customHTTP:
            return "\(name) · \(customHTTPMethod.uppercased()) \(customHTTPURL)"
        }
    }

    private static let defaultsKey = "AppSettings"
    private static let migratedSourceToSystemKey = "didMigrateSourceLanguageToSystem"
    private static let migratedDualLanguageKey = "didMigrateDualLanguagePairs"

    mutating func resetHotkeysToDefaults() {
        translateOnlyHotkey = .translateOnlyDefault
        translateAndSendHotkey = .translateAndSendDefault
        popupHotkey = .popupDefault
    }

    func save() {
        if let data = try? JSONEncoder().encode(CodableSettings(self)) {
            UserDefaults.standard.set(data, forKey: Self.defaultsKey)
        }
        Self.applyLoginItem(openAtLogin)
    }

    static func load() -> AppSettings {
        let settings: AppSettings
        if let data = UserDefaults.standard.data(forKey: defaultsKey),
           let decoded = try? JSONDecoder().decode(CodableSettings.self, from: data) {
            settings = decoded.asSettings()
            if ProviderKind(rawValue: decoded.providerKind) == nil {
                settings.save()
            }
        } else {
            settings = AppSettings()
        }
        return migrateDualLanguageIfNeeded(migrateSourceLanguageIfNeeded(settings))
    }

    /// One-time: old default was `nil` (auto). Switch outgoing source to the system language once.
    private static func migrateSourceLanguageIfNeeded(_ settings: AppSettings) -> AppSettings {
        guard !UserDefaults.standard.bool(forKey: migratedSourceToSystemKey) else {
            return settings
        }
        var migrated = settings
        if migrated.outgoingSourceLanguage == nil {
            migrated.outgoingSourceLanguage = LanguageCode.systemLanguage
            if migrated.outgoingTargetLanguage == migrated.outgoingSourceLanguage {
                migrated.outgoingTargetLanguage = LanguageCode.systemDefaultTarget
            }
            migrated.save()
        }
        UserDefaults.standard.set(true, forKey: migratedSourceToSystemKey)
        return migrated
    }

    /// One-time: seed incoming pair if this install only had a single source/target.
    private static func migrateDualLanguageIfNeeded(_ settings: AppSettings) -> AppSettings {
        guard !UserDefaults.standard.bool(forKey: migratedDualLanguageKey) else {
            return settings
        }
        var migrated = settings
        // Incoming defaults: detect → your language (typical “reading foreign chat”).
        if migrated.incomingTargetLanguage.isEmpty {
            migrated.incomingTargetLanguage = LanguageCode.systemLanguage
        }
        if migrated.incomingTargetLanguage == migrated.outgoingTargetLanguage,
           migrated.incomingSourceLanguage == nil,
           migrated.outgoingSourceLanguage != nil {
            // Keep outgoing as configured; incoming stays detect → system.
            migrated.incomingTargetLanguage = LanguageCode.systemLanguage
        }
        migrated.save()
        UserDefaults.standard.set(true, forKey: migratedDualLanguageKey)
        return migrated
    }

    static func applyLoginItem(_ enabled: Bool) {
        if #available(macOS 13.0, *) {
            do {
                if enabled {
                    try SMAppService.mainApp.register()
                } else {
                    try SMAppService.mainApp.unregister()
                }
            } catch {
                // Best-effort; ignore registration errors in v1.
            }
        }
    }
}

private struct CodableSettings: Codable {
    var isEnabled: Bool
    var enterTranslatesAndSends: Bool
    /// Legacy single-pair keys (pre dual-language).
    var targetLanguage: String?
    var sourceLanguage: String?
    var incomingSourceLanguage: String?
    var incomingTargetLanguage: String?
    var outgoingSourceLanguage: String?
    var outgoingTargetLanguage: String?
    /// Stored as string so unknown former cloud engines (groq, gemini, …) still decode.
    var providerKind: String
    var openAtLogin: Bool
    var translateOnlyHotkey: HotkeyChord?
    var translateAndSendHotkey: HotkeyChord?
    var popupModeEnabled: Bool?
    var popupHotkey: HotkeyChord?
    var showStatusHUD: Bool?
    var deeplUseFreeAPI: Bool
    var openAIBaseURL: String
    var openAIModel: String
    var customHTTPURL: String
    var customHTTPMethod: String
    var customHTTPHeadersJSON: String
    var customHTTPBodyTemplate: String
    var customHTTPResponsePath: String

    init(_ settings: AppSettings) {
        isEnabled = settings.isEnabled
        enterTranslatesAndSends = settings.enterTranslatesAndSends
        // Keep legacy keys in sync with outgoing for older builds / tooling.
        targetLanguage = settings.outgoingTargetLanguage
        sourceLanguage = settings.outgoingSourceLanguage
        incomingSourceLanguage = settings.incomingSourceLanguage
        incomingTargetLanguage = settings.incomingTargetLanguage
        outgoingSourceLanguage = settings.outgoingSourceLanguage
        outgoingTargetLanguage = settings.outgoingTargetLanguage
        providerKind = settings.providerKind.rawValue
        openAtLogin = settings.openAtLogin
        translateOnlyHotkey = settings.translateOnlyHotkey
        translateAndSendHotkey = settings.translateAndSendHotkey
        popupModeEnabled = settings.popupModeEnabled
        popupHotkey = settings.popupHotkey
        showStatusHUD = settings.showStatusHUD
        deeplUseFreeAPI = settings.deeplUseFreeAPI
        openAIBaseURL = settings.openAIBaseURL
        openAIModel = settings.openAIModel
        customHTTPURL = settings.customHTTPURL
        customHTTPMethod = settings.customHTTPMethod
        customHTTPHeadersJSON = settings.customHTTPHeadersJSON
        customHTTPBodyTemplate = settings.customHTTPBodyTemplate
        customHTTPResponsePath = settings.customHTTPResponsePath
    }

    func asSettings() -> AppSettings {
        var s = AppSettings()
        s.isEnabled = isEnabled
        s.enterTranslatesAndSends = enterTranslatesAndSends

        let legacySource = sourceLanguage
        let legacyTarget = targetLanguage ?? LanguageCode.systemDefaultTarget
        let hasDual =
            outgoingTargetLanguage != nil
            || incomingTargetLanguage != nil
            || outgoingSourceLanguage != nil
            || incomingSourceLanguage != nil

        if hasDual {
            s.outgoingTargetLanguage = outgoingTargetLanguage ?? legacyTarget
            // Explicit null means “Detectar idioma” — do not fall back to legacy.
            s.outgoingSourceLanguage = outgoingSourceLanguage
            s.incomingTargetLanguage = incomingTargetLanguage ?? LanguageCode.systemLanguage
            s.incomingSourceLanguage = incomingSourceLanguage
        } else {
            s.outgoingTargetLanguage = legacyTarget
            s.outgoingSourceLanguage = legacySource ?? LanguageCode.systemLanguage
            s.incomingTargetLanguage = LanguageCode.systemLanguage
            s.incomingSourceLanguage = nil
        }

        s.providerKind = ProviderKind(rawValue: providerKind) ?? .apple
        s.openAtLogin = openAtLogin
        s.translateOnlyHotkey = translateOnlyHotkey ?? .translateOnlyDefault
        s.translateAndSendHotkey = translateAndSendHotkey ?? .translateAndSendDefault
        s.popupModeEnabled = popupModeEnabled ?? false
        s.popupHotkey = popupHotkey ?? .popupDefault
        s.showStatusHUD = showStatusHUD ?? true
        s.deeplUseFreeAPI = deeplUseFreeAPI
        s.openAIBaseURL = openAIBaseURL
        s.openAIModel = openAIModel
        s.customHTTPURL = customHTTPURL
        s.customHTTPMethod = customHTTPMethod
        s.customHTTPHeadersJSON = customHTTPHeadersJSON
        s.customHTTPBodyTemplate = customHTTPBodyTemplate
        s.customHTTPResponsePath = customHTTPResponsePath
        return s
    }
}

File: Prism.Windows/src/Prism.Core/Settings/AppSettings.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Prism.Core.Hotkeys;
using Prism.Core.Translation;

namespace Prism.Core.Settings;

public sealed class AppSettings
{
    public bool IsEnabled { get; set; } = true;
    public bool EnterTranslatesAndSends { get; set; }

    public string? IncomingSourceLanguage { get; set; }
    public string IncomingTargetLanguage { get; set; } = LanguageCode.SystemLanguage;

    public string? OutgoingSourceLanguage { get; set; } = LanguageCode.SystemLanguage;
    public string OutgoingTargetLanguage { get; set; } = LanguageCode.SystemDefaultTarget;

    public ProviderKind ProviderKind { get; set; } = ProviderKind.Azure;
    public bool OpenAtLogin { get; set; }

    public HotkeyChord TranslateOnlyHotkey { get; set; } = HotkeyChord.TranslateOnlyDefault;
    public HotkeyChord TranslateAndSendHotkey { get; set; } = HotkeyChord.TranslateAndSendDefault;
    public bool PopupModeEnabled { get; set; }
    public HotkeyChord PopupHotkey { get; set; } = HotkeyChord.PopupDefault;
    public bool ShowStatusHud { get; set; } = true;

    public bool DeepLUseFreeApi { get; set; } = true;

    public string OpenAIBaseUrl { get; set; } = "http://localhost:11434/v1";
    public string OpenAIModel { get; set; } = "local-model";

    public string AzureRegion { get; set; } = "global";

    public string CustomHttpUrl { get; set; } = "http://localhost:8080/translate";
    public string CustomHttpMethod { get; set; } = "POST";
    public string CustomHttpHeadersJson { get; set; } = """{"Content-Type":"application/json"}""";
    public string CustomHttpBodyTemplate { get; set; } =
        """{"text":"{{text}}","target":"{{to}}","source":"{{from}}"}""";
    public string CustomHttpResponsePath { get; set; } = "translatedText";

    public string? ResolvedSourceLanguage(bool outgoing)
    {
        var raw = outgoing ? OutgoingSourceLanguage : IncomingSourceLanguage;
        return string.IsNullOrWhiteSpace(raw) ? null : raw;
    }

    public string TargetLanguage(bool outgoing) =>
        outgoing ? OutgoingTargetLanguage : IncomingTargetLanguage;

    public string EngineLogDescription => ProviderKind switch
    {
        ProviderKind.Azure => $"{ProviderKind.DisplayName()} · region {AzureRegion}",
        ProviderKind.DeepL => $"{ProviderKind.DisplayName()} · {(DeepLUseFreeApi ? "free API" : "Pro API")}",
        ProviderKind.Google => ProviderKind.DisplayName(),
        ProviderKind.OpenAICompatible =>
            $"{ProviderKind.DisplayName()} · model {OpenAIModel} · {OpenAIBaseUrl}",
        ProviderKind.CustomHttp =>
            $"{ProviderKind.DisplayName()} · {CustomHttpMethod.ToUpperInvariant()} {CustomHttpUrl}",
        _ => ProviderKind.DisplayName(),
    };

    public void ResetHotkeysToDefaults()
    {
        TranslateOnlyHotkey = HotkeyChord.TranslateOnlyDefault;
        TranslateAndSendHotkey = HotkeyChord.TranslateAndSendDefault;
        PopupHotkey = HotkeyChord.PopupDefault;
    }

    public string? ModelFor(ProviderKind kind) => kind switch
    {
        ProviderKind.OpenAICompatible => OpenAIModel,
        _ => null,
    };

    public void SetModel(string model, ProviderKind kind)
    {
        if (kind == ProviderKind.OpenAICompatible)
            OpenAIModel = model;
    }

    public static AppSettings CreateDefaults() => new();
}

public sealed class AppSettingsStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) },
    };

    private readonly string _path;

    public AppSettingsStore(string? path = null)
    {
        _path = path ?? DefaultPath();
    }

    public static string DefaultPath()
    {
        var root = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        return Path.Combine(root, "Prism", "settings.json");
    }

    public AppSettings Load()
    {
        try
        {
            if (!File.Exists(_path))
                return AppSettings.CreateDefaults();

            var json = File.ReadAllText(_path);
            var dto = JsonSerializer.Deserialize<SettingsDto>(json, JsonOptions);
            return dto?.ToSettings() ?? AppSettings.CreateDefaults();
        }
        catch
        {
            return AppSettings.CreateDefaults();
        }
    }

    public void Save(AppSettings settings)
    {
        var dir = Path.GetDirectoryName(_path);
        if (!string.IsNullOrEmpty(dir))
            Directory.CreateDirectory(dir);

        var dto = SettingsDto.From(settings);
        var json = JsonSerializer.Serialize(dto, JsonOptions);
        File.WriteAllText(_path, json);
    }

    private sealed class SettingsDto
    {
        public bool IsEnabled { get; set; } = true;
        public bool EnterTranslatesAndSends { get; set; }
        public string? IncomingSourceLanguage { get; set; }
        public string IncomingTargetLanguage { get; set; } = "en";
        public string? OutgoingSourceLanguage { get; set; }
        public string OutgoingTargetLanguage { get; set; } = "en";
        public string ProviderKind { get; set; } = "azure";
        public bool OpenAtLogin { get; set; }
        public HotkeyDto? TranslateOnlyHotkey { get; set; }
        public HotkeyDto? TranslateAndSendHotkey { get; set; }
        public bool PopupModeEnabled { get; set; }
        public HotkeyDto? PopupHotkey { get; set; }
        public bool ShowStatusHud { get; set; } = true;
        public bool DeepLUseFreeApi { get; set; } = true;
        public string OpenAIBaseUrl { get; set; } = "http://localhost:11434/v1";
        public string OpenAIModel { get; set; } = "local-model";
        public string AzureRegion { get; set; } = "global";
        public string CustomHttpUrl { get; set; } = "http://localhost:8080/translate";
        public string CustomHttpMethod { get; set; } = "POST";
        public string CustomHttpHeadersJson { get; set; } = """{"Content-Type":"application/json"}""";
        public string CustomHttpBodyTemplate { get; set; } =
            """{"text":"{{text}}","target":"{{to}}","source":"{{from}}"}""";
        public string CustomHttpResponsePath { get; set; } = "translatedText";

        public static SettingsDto From(AppSettings s) => new()
        {
            IsEnabled = s.IsEnabled,
            EnterTranslatesAndSends = s.EnterTranslatesAndSends,
            IncomingSourceLanguage = s.IncomingSourceLanguage,
            IncomingTargetLanguage = s.IncomingTargetLanguage,
            OutgoingSourceLanguage = s.OutgoingSourceLanguage,
            OutgoingTargetLanguage = s.OutgoingTargetLanguage,
            ProviderKind = s.ProviderKind.ToStorageKey(),
            OpenAtLogin = s.OpenAtLogin,
            TranslateOnlyHotkey = HotkeyDto.From(s.TranslateOnlyHotkey),
            TranslateAndSendHotkey = HotkeyDto.From(s.TranslateAndSendHotkey),
            PopupModeEnabled = s.PopupModeEnabled,
            PopupHotkey = HotkeyDto.From(s.PopupHotkey),
            ShowStatusHud = s.ShowStatusHud,
            DeepLUseFreeApi = s.DeepLUseFreeApi,
            OpenAIBaseUrl = s.OpenAIBaseUrl,
            OpenAIModel = s.OpenAIModel,
            AzureRegion = s.AzureRegion,
            CustomHttpUrl = s.CustomHttpUrl,
            CustomHttpMethod = s.CustomHttpMethod,
            CustomHttpHeadersJson = s.CustomHttpHeadersJson,
            CustomHttpBodyTemplate = s.CustomHttpBodyTemplate,
            CustomHttpResponsePath = s.CustomHttpResponsePath,
        };

        public AppSettings ToSettings() => new()
        {
            IsEnabled = IsEnabled,
            EnterTranslatesAndSends = EnterTranslatesAndSends,
            IncomingSourceLanguage = IncomingSourceLanguage,
            IncomingTargetLanguage = string.IsNullOrWhiteSpace(IncomingTargetLanguage)
                ? LanguageCode.SystemLanguage
                : IncomingTargetLanguage,
            OutgoingSourceLanguage = OutgoingSourceLanguage,
            OutgoingTargetLanguage = string.IsNullOrWhiteSpace(OutgoingTargetLanguage)
                ? LanguageCode.SystemDefaultTarget
                : OutgoingTargetLanguage,
            ProviderKind = ProviderKindExtensions.Parse(ProviderKind),
            OpenAtLogin = OpenAtLogin,
            TranslateOnlyHotkey = TranslateOnlyHotkey?.ToChord() ?? HotkeyChord.TranslateOnlyDefault,
            TranslateAndSendHotkey = TranslateAndSendHotkey?.ToChord() ?? HotkeyChord.TranslateAndSendDefault,
            PopupModeEnabled = PopupModeEnabled,
            PopupHotkey = PopupHotkey?.ToChord() ?? HotkeyChord.PopupDefault,
            ShowStatusHud = ShowStatusHud,
            DeepLUseFreeApi = DeepLUseFreeApi,
            OpenAIBaseUrl = OpenAIBaseUrl,
            OpenAIModel = OpenAIModel,
            AzureRegion = AzureRegion,
            CustomHttpUrl = CustomHttpUrl,
            CustomHttpMethod = CustomHttpMethod,
            CustomHttpHeadersJson = CustomHttpHeadersJson,
            CustomHttpBodyTemplate = CustomHttpBodyTemplate,
            CustomHttpResponsePath = CustomHttpResponsePath,
        };
    }

    private sealed class HotkeyDto
    {
        public int VirtualKey { get; set; }
        public byte Mods { get; set; }

        public static HotkeyDto From(HotkeyChord chord) => new()
        {
            VirtualKey = chord.VirtualKey,
            Mods = (byte)chord.Mods,
        };

        public HotkeyChord ToChord() => new()
        {
            VirtualKey = VirtualKey,
            Mods = (HotkeyChord.Modifiers)Mods,
        };
    }
}
